deps/spidershim/src/v8objecttemplate.cc

```diff
@@ -22,11 +22,430 @@
+  nullptr, // call
+  nullptr, // hasInstance
+  nullptr, // construct
+  nullptr  // trace
+};
+
+const JSClass objectTemplateClass = {
+  "ObjectTemplate",
+  JSCLASS_HAS_RESERVED_SLOTS(uint32_t(TemplateSlots::NumSlots)),
+  &objectTemplateClassOps
+};
+
+enum class InstanceSlots {
+  InstanceClassSlot,  // Stores a refcounted pointer to our InstanceClass*.
+  NumSlots
+};
+
+const JSClassOps objectInstanceClassOps = {
+  nullptr, // addProperty
+  nullptr, // delProperty
+  nullptr, // getProperty
+  nullptr, // setProperty
+  nullptr, // enumerate
+  nullptr, // resolve
+  nullptr, // mayResolve
+  ObjectTemplateFinalize<uint32_t(InstanceSlots::InstanceClassSlot)>,
+  nullptr, // call
+  nullptr, // hasInstance
+  nullptr, // construct
+  nullptr  // trace
+};
+
+enum class AccessorSlots {
+  NameSlot,
+  DataSlot,
+  CallbackSlot1,
+  CallbackSlot2,
+  NumSlots
+};
+
+const JSClass accessorDataClass = {
+  "AccessorData",
+  JSCLASS_HAS_RESERVED_SLOTS(uint32_t(AccessorSlots::NumSlots))
+};
+
+// Our getters/setters need to pass two things along to the underlying
+template<typename> struct CallbackTraits;
+
+template<>
+struct CallbackTraits<v8::AccessorGetterCallback> {
+  typedef v8::PropertyCallbackInfo<v8::Value> PropertyCallbackInfo;
+
+  static void doCall(v8::Isolate* isolate, v8::AccessorGetterCallback callback,
+                     v8::Local<v8::String> name,
+                     PropertyCallbackInfo& info, JS::CallArgs& args) {
+    callback(name, info);
+  }
+
+  static const unsigned nargs = 0;
+};
+
+template<>
+struct CallbackTraits<v8::AccessorSetterCallback> {
+  typedef v8::PropertyCallbackInfo<void> PropertyCallbackInfo;
+
+  static void doCall(v8::Isolate* isolate, v8::AccessorSetterCallback callback,
+                     v8::Local<v8::String> name, PropertyCallbackInfo& info,
+                     JS::CallArgs& args) {
+    using namespace v8;
+    Local<Value> value = internal::Local<Value>::New(isolate, args.get(0));
+
+    callback(name, value, info);
+  }
+
+  static const unsigned nargs = 0;
+};
+
+// AccessorGetterCallback/AccessorSetterCallback are passed information that is
+// not really present in SpiderMonkey: a Local<String> for the property name
+// (directly) and a Value for the callback data (indirectly, via
+// PropertyCallbackInfo).  We store those two things, and the actual address of
+// the V8 callback function, in the reserved slots of an object that hangs off
+// the accessor.
+template<typename CallbackType>
+bool NativeAccessorCallback(JSContext* cx, unsigned argc, JS::Value* vp) {
+  using namespace v8;
+  Isolate* isolate = Isolate::GetCurrent();
+  JS::CallArgs args = JS::CallArgsFromVp(argc, vp);
+  JS::RootedObject callee(cx, &args.callee());
+
+  // XXXbz This seems busted conceptually: "thisv" can be a non-object!  But
+  // objects are all PropertyCallbackInfo accepts....
+  Local<Object> thisObject =
+    internal::Local<Object>::New(isolate, args.thisv());
```

Comment:
Both here and in `NativeFunctionCallback`, we should check to make sure `thisv` is actually an object.